ipn/ipnlocal/local.go

```diff
@@ -1326,20 +1326,26 @@ func (b *LocalBackend) SetControlClientStatus(c controlclient.Client, st control
 			prefs.Persist = st.Persist.AsStruct()
 		}
 	}
+	if st.LoginFinished() {
```

Comment:
@zofrex can you review this bit too?

Reply:
It's a little hard to reason about, `wasBlocked` was standing in for both "we were running and we're doing a normal, non-seamless renewal" and "WantRunning was false", I think?

And `LoginFinished()` really just means "are we authenticated?" which doesn't necessarily mean a login finished on this tick.

This makes me wonder if the logic on line 1292 is correct, or needs amending too.

But yeah, this looks good to me 👍

Reply:
Re `LoginFinished()`, I think it does mean that a login has just finished. Well, kind of.

```go
const (
	StateNew = State(iota)
	StateNotAuthenticated
	StateAuthenticating
	StateURLVisitRequired
	StateAuthenticated
	StateSynchronized // connected and received map update
)
```
Both `StateAuthenticated` and `StateSynchronized` mean that we're authenticated. But `LoginFinished()` only reports `true` if the current state is `StateAuthenticated`.

So unless `controlclient` can report a status update after authentication but before receiving the first map update (which doesn't seem to be the case), then `LoginFinished()` reporting true indicates that authentication has just completed.

Reply:
Re `wasBlocked` and the logic on line 1292, I think it just means that engine updates were blocked on entry to `SetControlClientStatus`.

It seems we block engine updates when an interactive login is required. This includes "we're doing a normal, non-seamless renewal", "we don't have a node key and WantRunning is true" and "a user has initiated forced reauth".

It also seems that we might be using `wasBlocked` as a proxy for specific states, not just as an indicator that engine updates are blocked, which I believe is error-prone.

But the logic on line 1292 makes sense to me.